[Layout foundations] Add alpha `Box` component

laurkim opened the pull request:
### WHY are these changes introduced?

Resolves #6892. 

<!--
  Context about the problem that’s being addressed.
-->

### WHAT is this pull request doing?

Adds a new alpha `Box` component.
    <details>
      <summary>Using `Box` to recreate the `Card` component</summary>
      <img src="https://user-images.githubusercontent.com/26749317/187263883-0e8194f5-fbd8-4227-bdf5-66e868ffa573.png" alt="Using `Box` to recreate the `Card` component">
    </details>

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<!--
  Give as much information as needed to experiment with the component
  in the playground.
-->

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';

import {Page, Card, Text, Box} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
      <Box
        background="surface"
        borderRadius="radius-2"
        padding="5"
        shadow="card"
      >
        <Text as="h2" variant="headingMd">
          Online store dashboard
        </Text>
        <Text as="p" variant="bodyMd">
          This was made using the 📦 Box 📦 component.
        </Text>
      </Box>
      <div
        style={{
          borderTop: 'var(--p-border-divider)',
          marginTop: 'var(--p-space-8)',
          marginBottom: 'var(--p-space-8)',
        }}
      ></div>
      <Card title="Online store dashboard" sectioned>
        <Text as="p" variant="bodyMd">
          This was made using the 🃏 Card 🃏 component.
        </Text>
      </Card>
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

laurkim committed:
[Layout foundations] Add `react-polymorphic` dep

laurkim committed:
[Layout foundations] Add polymorphic box component

laurkim committed:
[Layout foundations] Add `background` prop

laurkim committed:
[Layout foundations] Add `borderRadius` prop

laurkim committed:
[Layout foundations] Add `shadow` prop

laurkim committed:
Add changeset

laurkim committed:
Fix merge conflict

laurkim committed:
[Layout foundations] Add `margin` and `padding` props

Co-authored-by: Aaron Casanova <aaron.casanova@shopify.com>

alex-page commented on polaris-react/src/components/Box/Box.scss:
```diff
@@ -0,0 +1,172 @@
+  --pc-box-padding-top: initial;
+
+  display: block;
+  margin: var(--pc-box-margin);
```
We could remove the above lines of code by saying if the custom property isn't defined then it falls back to `initial`.

```suggestion
  margin: var(--pc-box-margin, initial);
```

laurkim committed:
[Layout foundations] Refactor `margin` and `padding` logic

Including the `margin/padding` property was causing overrides when combining with `marginTop` or `paddingLeft`, for example.

Co-authored-by: Aveline Thelen <aveline.thelen@shopify.com>
Co-authored-by: Alex Page <alex.page@shopify.com>

laurkim committed:
[Layout foundations] Refactor `borderRadius` to use `shape` tokens

laurkim committed:
[Layout foundations] Refactor and add `border` and `borderRadius` props

laurkim committed:
[Layout foundations] Refactor `shadow` to use `depth` tokens

alex-page commented on polaris-react/src/components/Box/Box.tsx:
```diff
@@ -0,0 +1,263 @@
+  shadow?: DepthTokenScale;
+}
+
+type PolymorphicBox = Polymorphic.ForwardRefComponent<'div', BoxBaseProps>;
```
Would we want to start with a more opinionated list? Then add flexibility later?

laurkim replied:
I originally tried building `Box` out that way but was running into a lot of TS errors. I can revisit and see if I can get it working without `@radix-ui/react-polymorphic` package if we prefer that approach?

alex-page replied:
I think initially if we can get this working without allowing all the HTML elements that would be ace.

kyledurand commented on polaris-react/src/components/Box/Box.tsx:
```diff
@@ -0,0 +1,263 @@
+
+type BorderRadiusTokenScale = Extract<
+  BorderShapeTokenScale,
+  `radius-${string}`
```
Spacing takes a numerical string without prefix but border radius requires one:
```ts
padding="1"
borderRadius="radius-4"
```

Can we make it consistent? I'm fine with `space-1` or `borderRadius="4"` Whatever you think makes the most sense.

laurkim replied:
I'll update `borderRadius` to remove the prefix and be typed as `base` | `05` | `1` | etc.

alex-page replied:
What is `base`? I always disliked `base` as a value...

laurkim replied:
I included `border-base`, `border-dark`, `border-transparent`, `border-divider`, and `border-divider-on-dark` from our `shape` tokens. We have a `border-base` [token](https://github.com/Shopify/polaris/blob/7a548a00a9b723e3e074801725a0361a4ec888f8/polaris-tokens/src/token-groups/shape.ts#L47) and since I'm using TS to remove the `border-` prefix, we're left with just `base`.

alex-page replied:
Right LOL. I think that is a token naming convention we should/could challenge as it's not clear to the user what it does.

laurkim replied:
100%, I know we had previous discussions about modifying some of the naming structure for tokens parallel to the layout foundations work so that can be included there as well 👍

kyledurand commented on polaris-react/src/components/Box/Box.tsx:
```diff
@@ -0,0 +1,263 @@
+      '--pc-box-shadow': shadow ? `var(--p-shadow-${shadow})` : '',
+    } as React.CSSProperties;
+
+    const className = classNames(styles.root, background && styles[background]);
```
`background && styles[background]`

Can we just assign the background color to a custom property instead of creating classes for each?

alex-page replied:
I like this as it allows us to iterate without adding more CSS.

kyledurand reviewed (commented):
This looking good so far! One other thing I noticed is that I can make a box with a dark background but can't change the text color. So the text gets lost. Can that be addressed within the Text component? (I haven't looked at it recently)

kyledurand commented on polaris-react/src/components/Box/Box.scss:
```diff
@@ -0,0 +1,114 @@
+.root {
```
Is this a new pattern we want to adopt? Typically the root class is named after the component

```suggestion
.Box {
```

laurkim committed:
[Layout foundations] Remove `radius-` prefix from `borderRadius` type

laurkim committed:
[Layout foundations] Refactor `background` to use color tokens

laurkim committed:
[Layout foundations] Update `root` classname to `Box`

Co-authored-by: Kyle Durand <kyledurand@users.noreply.github.com>

alex-page reviewed (approved):
Happy to ship and iterate in the `layout-foundations-prototype` branch @laurkim
